lib/controllers/hosted-skill-controller/helper.js

```diff
@@ -81,3 +95,66 @@ function handleSkillStatus(response, skillId, callback) {
         callback(`${errMessage}${errReason}Infrastructure provision for the hosted skill failed. Please try again.`);
     }
 }
+
+function downloadAuthInfoScript(callback) {
+    const authInfoUrl = CONSTANTS.HOSTED_SKILL.S3_SCRIPTS.AUTH_INFO.URL;
+    const authInfoPath = path.join(os.homedir(),
+        CONSTANTS.FILE_PATH.ASK.HIDDEN_FOLDER,
+        CONSTANTS.FILE_PATH.ASK.AUTH_INFO);
+    this.downloadScriptFromS3(authInfoUrl, authInfoPath, (err) => {
+        if (err) {
+            return callback(err);
+        }
+        return callback(null);
+    });
+}
+
+function downloadAskPrePushScript(callback) {
+    const askScriptUrl = CONSTANTS.HOSTED_SKILL.S3_SCRIPTS.ASK_PRE_PUSH.URL;
+    const scriptFolderPath = path.join(os.homedir(),
+        CONSTANTS.FILE_PATH.ASK.HIDDEN_FOLDER,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.NAME);
+    const askFilePath = path.join(os.homedir(),
+        CONSTANTS.FILE_PATH.ASK.HIDDEN_FOLDER,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.NAME,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.ASK_PRE_PUSH);
+    if (!fs.existsSync(scriptFolderPath)) {
+        fs.mkdirSync(scriptFolderPath);
+    }
+    this.downloadScriptFromS3(askScriptUrl, askFilePath, (err) => callback(err || null));
+}
+
+function downloadGitCredentialHelperScript(callback) {
+    const credentialScriptUrl = CONSTANTS.HOSTED_SKILL.S3_SCRIPTS.GIT_CREDENTIAL_HELPER.URL;
+    const scriptFolderPath = path.join(os.homedir(),
+        CONSTANTS.FILE_PATH.ASK.HIDDEN_FOLDER,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.NAME);
+    const credentialFilePath = path.join(os.homedir(),
+        CONSTANTS.FILE_PATH.ASK.HIDDEN_FOLDER,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.NAME,
+        CONSTANTS.FILE_PATH.ASK.SCRIPTS_FOLDER.GIT_CREDENTIAL_HELPER);
+    if (!fs.existsSync(scriptFolderPath)) {
+        fs.mkdirSync(scriptFolderPath);
+    }
+    this.downloadScriptFromS3(credentialScriptUrl, credentialFilePath, (err) => callback(err || null));
+}
+
+function downloadScriptFromS3(scriptUrl, filePath, callback) {
+    const DOWNLOADED = 'downloaded';
+    const retryCall = (loopCallback) => {
+        const request = https.get(scriptUrl, (res) => {
+            const writeStream = fs.createWriteStream(filePath);
+            res.pipe(writeStream);
+            res.on('error', (err) => {
+                loopCallback(new Error(err));
```

Comment:
I think this would cause a racing condition. You should resolve callback when you are done writing to the stream(file) not downloading the file.

example:
var http = require('http');
var fs = require('fs');

```
var download = function(url, dest, cb) {
  var file = fs.createWriteStream(dest);
  var request = http.get(url, function(response) {
    response.pipe(file);
    file.on('finish', function() {
      file.close(cb);  // close() is async, call cb after close completes.
    });
  }).on('error', function(err) { // Handle errors
    fs.unlink(dest); // Delete the file async. (But we don't check the result)
    if (cb) cb(err.message);
  });
};
```

https://stackoverflow.com/questions/11944932/how-to-download-a-file-with-node-js-without-using-third-party-libraries

May be also extract download to its own method. Probably can be reused.

Reply:
You are right. 
To move `const writeStream = fs.createWriteStream(filePath);` out of the request block seems also resolve the problem.